Type generics

FichteFoll opened the pull request:
This ~~is not 100% accurate but close enough and~~ resolves all basedpyright errors (with the default config).

I totally should not be doing this right now, but my body …!

Also note that I did not actually test whether this is valid in Python 3.8, but considering most of the complex typing stuff is behind the `TYPE_CHECKING` check, I believe it should work. Unfortunately, the ParamSpec syntax of 3.12 cannot be used.

FichteFoll committed:
Add generic type annotations

This is not 100% accurate but close enough and resolves all basedpyright
errors (with the default config).

FichteFoll committed:
Address some basedpyright warnings

There are *many* more that I believe are not worth addressing (at least
not right now).

FichteFoll commented:
The decorator annotation doesn't quite work yet. Needs more care.
(Can I not turn an already opened pull request into a draft again?)

FichteFoll commented:
Okay, I got a version that works as expected for basedpyright, but I have not checked pyright or mypy. Notably mypy is likely not going to work because I rely on type inference on the `debounce` decorator since I couldn't get the signature of the overloaded sub-decorator to work as the return value.

deathaxe commented on src/sublime_aio.py:
```diff
@@ -1,20 +1,36 @@
+        | None
+    )
+
+    EL = TypeVar('EL', bound=sublime_plugin.EventListener)
+    VEL = TypeVar('VEL', bound=sublime_plugin.ViewEventListener)
```
As we use those vars in decorator function which works only within `sublime_aio.EventListeners`, shouldn't those vars point to the more specific classes then?

```suggestion
    EL = TypeVar('EL', bound=EventListener)
    VEL = TypeVar('VEL', bound=ViewEventListener)
```

This way we get feedback of misuse...

![grafik](https://github.com/user-attachments/assets/cc0f844e-09bd-445e-a123-7ec8e4dd3c7b)


_Note: I find it odd those to be needed just to avoid pyright not to throw typing errors. I'd expect following `@overload` to also work when directly using `EventListener` types._

FichteFoll replied:
Good point. Adopted this.

deathaxe commented on src/sublime_aio.py:
```diff
@@ -58,12 +76,12 @@ def shutdown():
-def debounced(delay_in_ms: int) -> Callable:
-    """Schedule coroutine for execution as soon as no events arive for given delay.
+
+def debounced(delay_in_ms: int):
```
Isn't the returned `wrapper` a `Callable`?

FichteFoll replied:
Yes, but I couldn't get the signature right so I relied on pyright's type inference here. Will add a comment about it.

deathaxe commented on src/sublime_aio.py:
```diff
@@ -106,13 +132,13 @@ def _debounced_callback(view: sublime.View, coro_args: tuple) -> None:
-                    coro_func, self, args, kwargs = coro_args
-                    __loop.create_task(coro_func(self, *args, **kwargs))
+                    # Tasks are collected later using `asyncio.all_tasks`.
+                    _ = __loop.create_task(coro_func())
```
Why assigning a nonused return value to just nothing? Isn't that just extra noise?

deathaxe replied:
Disabling all filters of basedpyright, I see the warning without `_ = `, but this is one of the runtime affecting messages, I find not worth it. Not only adding un-pythonic noise to the code it also executes useless assignment operation during runtime. Same as calling `cast()` shit.

FichteFoll replied:
That was indeed a basedpyright warning, which has a couple functions' return values defined as "must use". The `_ =` syntax can be used to explicitly opt out of it.

I generally believe it makes sense to have this check, but it seemed a bit invasive on this code since there are quite a lot of instances with asynchronous computing where Python generally wants you to do something with the return value, mostly to cancel them.

I only added it in this particular case because it seemed important enough to even mention it in a comment. After switching to the `base` type checking mode, these warnings also don't show up anymore.

deathaxe commented on src/sublime_aio.py:
```diff
@@ -106,13 +132,13 @@ def _debounced_callback(view: sublime.View, coro_args: tuple) -> None:
+            __loop.call_at(call_at[view.view_id], _debounced_callback, view, coro_func)
 
-        def wrapper(self, *args, **kwargs) -> None:
+        def wrapper(self: sublime_plugin.EventListener | sublime_plugin.ViewEventListener, *args: sublime.View) -> None:
```
I'd proably also point to `sublime_aio` Listeners here.

```suggestion
        def wrapper(self: EventListener | ViewEventListener, *args: sublime.View) -> None:
```

Also assigning `*args` the `sublime.View` type feels odd as it is possibly at most only the first argument, which is of that type.

FichteFoll replied:
`sublime_aio` listeners have been adopted.

The `*args: sublime.View` was just the only idea I had to get the signature to capture both the EventListener and the ViewEventListener case with that argument that is only present in one case. I'm open for different ideas, but for now I find this acceptable.

deathaxe replied:
Just wondered it's working.

deathaxe commented on src/sublime_aio.py:
```diff
@@ -124,21 +150,21 @@ def wrapper(self, *args, **kwargs) -> None:
                 return
 
-            view = self.view if hasattr(self, 'view') else args[0]
+            view = self.view if isinstance(self, sublime_plugin.ViewEventListener) else args[0]
```
As decorator is valid only for our own listeners...

```suggestion
            view = self.view if isinstance(self, ViewEventListener) else args[0]
```

deathaxe commented on src/sublime_aio.py:
```diff
@@ -1,20 +1,36 @@
+    T = TypeVar('T')
+
+    BlankCoro: TypeAlias = Coroutine[object, None, None]
+    CompletionsReturnVal: TypeAlias = (
+        sublime.CompletionList
+        | tuple[list[sublime.CompletionValue], sublime.AutoCompleteFlags]
+        | list[sublime.CompletionValue]
+        | None
+    )
```
This is what `sublime_types.CompletionsListValue` should be used for, but it doesn't seem to be as correct as this one. Maybe we suggest a fix.

FichteFoll replied:
I would agree, but `sublime_types` is not part of the LSP stubs and the `CompletionValue` that they define is not exported, so I resorted to just re-defining it.

I also don't have a `CompletionsListValue` in `sublime_types` on my machine.

Edit: Actually, it doesn't even work like this. I have corrected this type.

Edit2: Actually actually, my previous edit was based on the LSP type stubs, but they may be incorrect. I'll have to test with ST directly whether list item types can be mixed.

deathaxe replied:
Yeah, actually this is something I've defined downstream in my patched versions of ST APIs. The comment rather targeted into direction of proposing to add it to sublime_types and export it as various listeners make use of it.

deathaxe commented on src/sublime_aio.py:
```diff
@@ -87,10 +105,18 @@ async def on_modified(self):
-    def decorator(coro_func: Coroutine) -> Callable:
-        call_at = {}
+    @overload
+    def decorator(coro_func: Callable[[EL, sublime.View], BlankCoro]) -> Callable[[EL, sublime.View], None]: ...
```
Not sure how clear it is for endusers `EL` pointing to `sublime_aio.EventListener` and `VEL` to `sublime_aio.ViewEventListener` assuming they take the time to analyze and understand the messy output at all.

![grafik](https://github.com/user-attachments/assets/c698f049-b16e-481d-b984-829705f76ed3)

FichteFoll replied:
Unfortunately, we need these typevars.

This is with the typevars replaced with type aliases for the actual class types because it cannot resolve `self` properly:

![image](https://github.com/user-attachments/assets/020cecd8-39cd-40a2-95ee-21408c6a4afc)

deathaxe commented on src/sublime_aio.py:
```diff
@@ -292,17 +318,18 @@ class AsyncEventListenerMeta(type):
     """
 
-    def __new__(mcs, name, bases, attrs):
+    def __new__(mcs: type[AsyncEventListenerMeta], name: str, bases: tuple[type, ...], attrs: dict[str, object]) -> AsyncEventListenerMeta:
```
Does it make sense to wrap those long lines? This one even exceeds 120 chars limit.

```py
    def __new__(
        mcs: type[AsyncEventListenerMeta],
        name: str,
        bases: tuple[type, ...],
        attrs: dict[str, object]
    ) -> AsyncEventListenerMeta:
```

One arg per line also helps with readability, IMHO, especially if type hints are more complex.

FichteFoll replied:
Yes, totally. I just forgot about this, probably.

FichteFoll committed:
Add basic pyright configuration

FichteFoll committed:
Make type annotations compatible with Python 3.8

Even though the `if TYPE_CHECKING:` block should not be evaluated, tools
still complain about usage of 3.9 features such as `|` unions and basic
type subscripting.

FichteFoll committed:
Add typing-extensions as a dev dependency

FichteFoll committed:
Restrict type variables to our event listener types

FichteFoll committed:
Fix usage of `CompletionValue`

The values cannot be list within a list, so the type alias must be a
union of different lists.

FichteFoll committed:
Add a comment about type inference

FichteFoll committed:
Split long function definitions

deathaxe commented on src/sublime_aio.py:
```diff
@@ -1,20 +1,40 @@
+    T = TypeVar('T')
+
+    BlankCoro: TypeAlias = Coroutine[object, None, None]
+    CompletionsList: TypeAlias = Union[List[str], List[Tuple[str, str]], List[sublime.CompletionItem]]
+    CompletionsReturnVal: TypeAlias = Union[
+        sublime.CompletionList,
+        Tuple[CompletionsList, sublime.AutoCompleteFlags],
+        CompletionsList,
+        None,
+    ]
```
This one is causing type match errors as API expects `list[CompletionValue]`. It can be used here as `CompletionValue` is exported by `sublime` module in type checking mode.

Furthermore, 

1. as there's already a `sublime.CompletionList`, I'd avoid defining a type alias with nearly equal name.

2. a completion list can consist of items of mixed types (string, tuple and CompletionItem).

Hence the previous version was more correct.

FichteFoll replied:
The previous version does not validate against LSP stubs, however. We'll need to decide which is more correct, ideally based on what ST actually supports, and adjust our own stubs (to eventually be pushed as a wheel to pypi or github) accordingly.

https://github.com/sublimelsp/LSP/blob/177ae9b25c8242e697fc96767cf32363329958f1/stubs/sublime.pyi#L3135

deathaxe replied:
I'd rather follow official API and update stubs. Not sure when they have been adjusted to upstream changes.

deathaxe replied:
LSPs stubs are probably what plugins would do, but those are incorrect with regards to what ST API supports/expects. Not sure how LSP leverages those stubs, but I see type errors when using this form.

deathaxe committed:
Fix CompletionsReturnValue

Reverts 8e5c361e5ff7a26d89a6d4473059985e8619acaa
and uses `sublime.CompletionValue` to fix type mismatch with existing ST API
method `CompletionList.set_completions()`.

deathaxe committed:
Import only partial from functools

deathaxe committed:
Remove unused return value assignment
